### DISCUSSION TITLE
Flag "self-signed" in root certificate is missing when switching to new swanctl style configuration

### DISCUSSION THREAD

**Discussion by stephanhendl** [strongswan/strongswan#1846] [General]
Hello all,

we've a self-signed root CA with a corresponding root certificate, generated bei openssl several years ago. This root CA has signed our issuing CA from which all the client certificates are created. 

Running strongswan-5.9.5-2 on Ubuntu-22.04 (LTS)  with the ipsec configuration the root certificate was marked with `flags: CA CRLSign self-signed` which worked as expected. 

```
swanctl --list-certs --type x509 --flag ca --subject "C=DE, ST=Brandenburg, L=Potsdam, O=Landtag Brandenburg, OU=Verwaltung, CN=Landtag Brandenburg UCS Root CA, E=ca@landtag.brandenburg.de"

List of X.509 CA Certificates

  subject:  "C=DE, ST=Brandenburg, L=Potsdam, O=Landtag Brandenburg, OU=Verwaltung, CN=Landtag Brandenburg UCS Root CA, E=ca@landtag.brandenburg.de"
  issuer:   "C=DE, ST=Brandenburg, L=Potsdam, O=Landtag Brandenburg, OU=Verwaltung, CN=Landtag Brandenburg UCS Root CA, E=ca@landtag.brandenburg.de"
  validity:  not before Feb 18 16:58:21 2011, ok
             not after  Feb 18 16:58:21 2036, ok (expires in 4574 days)
  serial:    82:a6:9a:fe:9e:59:12:07
  altNames:  ca@landtag.brandenburg.de
  flags:     CA CRLSign self-signed
  authkeyId: 56:3f:82:b6:16:c6:bf:46:18:de:6c:b3:a0:e1:c4:10:52:82:9d:41
  subjkeyId: 56:3f:82:b6:16:c6:bf:46:18:de:6c:b3:a0:e1:c4:10:52:82:9d:41
  pubkey:    RSA 2048 bits
  keyid:     00:c4:c5:97:27:a2:71:e6:31:f9:1d:71:bd:5a:05:7d:df:32:97:0a
  subjkey:   56:3f:82:b6:16:c6:bf:46:18:de:6c:b3:a0:e1:c4:10:52:82:9d:41
```

Running strongswan-5.9.5-3.el9.x86_64 or strongswan-5.9.10 on CentOS 9 Stream with the new swanctl configuration the flag "self-signed" `flags: CA CRLSign` is missing.

```
swanctl --list-certs --type x509 --flag ca --subject "C=DE, ST=Brandenburg, L=Potsdam, O=Landtag Brandenburg, OU=Verwaltung, CN=Landtag Brandenburg UCS Root CA, E=ca@landtag.brandenburg.de"

List of X.509 CA Certificates

  subject:  "C=DE, ST=Brandenburg, L=Potsdam, O=Landtag Brandenburg, OU=Verwaltung, CN=Landtag Brandenburg UCS Root CA, E=ca@landtag.brandenburg.de"
  issuer:   "C=DE, ST=Brandenburg, L=Potsdam, O=Landtag Brandenburg, OU=Verwaltung, CN=Landtag Brandenburg UCS Root CA, E=ca@landtag.brandenburg.de"
  validity:  not before Feb 18 16:58:21 2011, ok
             not after  Feb 18 16:58:21 2036, ok (expires in 4574 days)
  serial:    82:a6:9a:fe:9e:59:12:07
  altNames:  ca@landtag.brandenburg.de
  flags:     CA CRLSign
  authkeyId: 56:3f:82:b6:16:c6:bf:46:18:de:6c:b3:a0:e1:c4:10:52:82:9d:41
  subjkeyId: 56:3f:82:b6:16:c6:bf:46:18:de:6c:b3:a0:e1:c4:10:52:82:9d:41
  pubkey:    RSA 2048 bits
  keyid:     00:c4:c5:97:27:a2:71:e6:31:f9:1d:71:bd:5a:05:7d:df:32:97:0a
  subjkey:   56:3f:82:b6:16:c6:bf:46:18:de:6c:b3:a0:e1:c4:10:52:82:9d:41
```

Here you can find the contents of the certificate:

```
openssl x509 -in LandtagBrandenburgUCSRootCA.crt -noout -text
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            82:a6:9a:fe:9e:59:12:07
        Signature Algorithm: sha1WithRSAEncryption
        Issuer: C = DE, ST = Brandenburg, L = Potsdam, O = Landtag Brandenburg, OU = Verwaltung, CN = Landtag Brandenburg UCS Root CA, emailAddress = ca@landtag.brandenburg.de
        Validity
            Not Before: Feb 18 15:58:21 2011 GMT
            Not After : Feb 18 15:58:21 2036 GMT
        Subject: C = DE, ST = Brandenburg, L = Potsdam, O = Landtag Brandenburg, OU = Verwaltung, CN = Landtag Brandenburg UCS Root CA, emailAddress = ca@landtag.brandenburg.de
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:bd:92:33:96:5c:f7:30:7e:1a:48:1a:dc:f3:90:
                    3b:bc:78:c5:e1:c5:35:98:76:43:c1:8b:25:5f:eb:
                    74:d3:0a:a0:1a:21:ca:1d:17:4e:bc:0f:3e:65:08:
                    7b:29:1f:38:9d:59:bb:da:1d:f3:ad:3d:6a:02:94:
                    66:06:c0:8d:81:f7:61:c3:1e:2d:5a:16:be:96:8d:
                    4b:5b:54:92:88:2a:4a:14:34:d7:05:d5:f2:47:c5:
                    ab:f1:3c:82:b9:5d:e1:75:23:36:9e:e5:1a:54:cd:
                    27:eb:bb:a8:14:94:7e:10:d4:74:f3:77:85:38:cd:
                    64:19:b6:af:e2:9f:75:59:1f:70:06:f8:9e:ca:10:
                    20:8a:df:ec:2e:20:da:4e:b7:ce:9d:6a:71:03:42:
                    90:e7:73:c2:1b:f1:5d:4b:f9:37:b6:bd:3e:f2:34:
                    c2:c4:d9:67:64:3c:c6:93:79:72:68:48:49:c8:0e:
                    6d:fe:4e:5c:0a:92:eb:18:f1:93:e7:fe:e6:60:8b:
                    92:b0:37:75:0c:dd:46:df:2a:72:fb:92:e2:47:d9:
                    2e:88:93:66:d6:5d:38:e7:94:f7:e1:61:5b:a3:44:
                    c0:c1:86:06:ed:fd:7a:bc:c4:cc:02:92:1a:51:87:
                    68:97:44:f6:e1:02:41:41:dd:c6:d4:e0:30:fd:a9:
                    03:89
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Basic Constraints: critical
                CA:TRUE
            X509v3 Subject Key Identifier:
                56:3F:82:B6:16:C6:BF:46:18:DE:6C:B3:A0:E1:C4:10:52:82:9D:41
            X509v3 Authority Key Identifier:
                keyid:56:3F:82:B6:16:C6:BF:46:18:DE:6C:B3:A0:E1:C4:10:52:82:9D:41
                DirName:/C=DE/ST=Brandenburg/L=Potsdam/O=Landtag Brandenburg/OU=Verwaltung/CN=Landtag Brandenburg UCS Root CA/emailAddress=ca@landtag.brandenburg.de
                serial:82:A6:9A:FE:9E:59:12:07
            X509v3 Key Usage:
                Certificate Sign, CRL Sign
            Netscape Cert Type:
                SSL CA, S/MIME CA, Object Signing CA
            X509v3 Subject Alternative Name:
                email:ca@landtag.brandenburg.de
            X509v3 Issuer Alternative Name:
                email:ca@landtag.brandenburg.de
            Netscape Comment:
                This certificate is a Root CA Certificate
    Signature Algorithm: sha1WithRSAEncryption
    Signature Value:
        1b:a3:7c:f4:3e:03:25:da:f2:d2:cf:60:18:bc:e9:1e:5b:c8:
        61:fd:e9:ab:bf:4c:37:86:8f:a6:f4:80:22:44:31:b1:18:ce:
        b3:3b:a1:d1:cf:ab:be:42:3a:bf:13:be:f9:a4:c9:2a:4d:28:
        40:d5:71:0f:07:67:a9:df:0f:0f:84:63:b3:a7:d4:c2:1f:bd:
        bf:1f:e8:b3:13:ec:14:63:16:c1:56:b8:82:43:1c:a1:c1:8d:
        b2:9d:ad:56:de:1a:14:cf:d0:2b:48:f0:b3:75:f0:2d:6f:c5:
        10:2b:60:cb:59:72:61:29:4b:62:3f:c7:39:33:56:53:74:0d:
        59:56:d8:d8:b2:ef:3d:fa:47:09:f4:6c:23:73:2b:de:38:ca:
        49:ea:2f:cd:a4:2a:81:3d:47:eb:2d:08:2d:5a:30:c7:06:b3:
        e5:21:0e:bf:3f:cb:37:15:44:0a:4f:b9:00:e4:fe:ad:38:fb:
        72:4b:19:e9:31:a6:9c:29:6e:ed:9e:fd:2f:b1:c4:5b:e1:66:
        69:5c:36:9d:aa:42:b1:e8:a6:25:c8:bb:2d:a4:06:bc:e7:0d:
        d3:3d:a5:46:28:64:56:28:cf:bd:db:08:31:f6:dc:d7:fc:04:
        58:b9:d1:80:e3:48:f2:58:78:64:d2:68:52:d4:a8:ad:ad:40:
        73:4b:9e:e3
```

That leads to errors like:

```
Aug 11 09:50:42 server charon-systemd[89144]: no issuer certificate found for "C=DE, ST=Brandenburg, L=Potsdam, O=Landtag Brandenburg, OU=Verwaltung, CN=Landtag Brandenburg UCS Root CA, E=ca@landtag.brandenburg.de"
...
Aug 11 09:50:42 server charon-systemd[89144]: no trusted RSA public key found for 'Client'

```

What's missing or misconfigured? 

Best regards
Stephan

**Comment by strongX509** [comment]
I cannot really fathom why the **`self-signed`** flag went missing with strongSwan 5.9.5 on CentOS 9. You might try to increase the debug levels in **`strongswan.conf`** to
 ```
syslog {
   daemon {
      asn = 3
      lib = 3
  }
}
```
or with with **`systemd`**
```
journal {
   asn = 3
   lib = 3
}
```
which might return some information on the certificate parsing.

**Comment by stephanhendl** [comment]
I've set the debug levels, restarted and tried to establish a connection with an iPad named `VTB19001.edp.ipad` in the SAN entry of th certificate. Logfile is enclosed.
[20230814_logging_cert.txt](https://github.com/strongswan/strongswan/files/12331727/20230814_logging_cert.txt)

Unfortunately I havn't seen neither errors nor the word "self_signed" in the debug output.

**Comment by tobiasbrunner** [comment]
Your subject and authority key identifiers don't match, which isn't compliant with RFC 5280. Since 5.9.3, if both identifiers are available and they don't match, the tested certificate is not considered issued by the issuer certificate (for the root CA cert that's the same certificate).

**Comment by stephanhendl** [comment]
Thank you for providing the reason @tobiasbrunner. But unfortuantely I cannot see where is the difference:

```
            X509v3 Subject Key Identifier:
                56:3F:82:B6:16:C6:BF:46:18:DE:6C:B3:A0:E1:C4:10:52:82:9D:41
            X509v3 Authority Key Identifier:
                keyid:56:3F:82:B6:16:C6:BF:46:18:DE:6C:B3:A0:E1:C4:10:52:82:9D:41
```

Both values seems to be identical (except of the keyword `keyid`).

Also the `swanctl --list-certs --type x509 --flag ca` command (see initial comment) shows identical values:

```
  authkeyId: 56:3f:82:b6:16:c6:bf:46:18:de:6c:b3:a0:e1:c4:10:52:82:9d:41
  subjkeyId: 56:3f:82:b6:16:c6:bf:46:18:de:6c:b3:a0:e1:c4:10:52:82:9d:41
```

**Reply by tobiasbrunner** [reply]
Sorry about that. You're right, or course. I wrote that comment after looking at the sub CA in the log and incorrectly interpreting that as root CA cert.

I've extracted the certs from the log and was able to successfully verify that they are valid:
```
$ pki --verify --cacert sub.der --cacert root.der < peer.der
  using certificate "DC=de, DC=lvnbb, DC=ltbbg1, OU=Verwaltung, OU=tablets, CN=VTB19001"
  using trusted intermediate ca certificate "CN=Landtag Brandenburg AD Sub CA 1"
  using trusted ca certificate "C=DE, ST=Brandenburg, L=Potsdam, O=Landtag Brandenburg, OU=Verwaltung, CN=Landtag Brandenburg UCS Root CA, E=ca@landtag.brandenburg.de"
  reached self-signed root ca with a path length of 1
certificate trusted, lifetimes valid
```

So what happens here is quite weird. But I think I have an explanation.

Looking at the plugin order we see this:
```
Aug 14 07:25:18 vpn4 charon-systemd[539242]: loaded plugins: charon-systemd ... x509 ... openssl ... gmp ...
```
So while the certificates are parsed by the _x509_ plugin and not the _openssl_ plugin, the RSA implementation is provided by the latter and not the _gmp_ plugin.

This is relevant because OpenSSL on newer Fedora/CentOS/RH systems has restrictions regarding RSA signatures with SHA-1. And the latter is exactly the signature scheme used to self-sign your root CA certificate (the sub CA and the peer certificate are both signed with SHA-256, so they are fine). So while the certificate can be parsed perfectly fine, the signature in it can't be verified by the RSA key implementation provided by the _openssl_ plugin via OpenSSL. That's why the certificate is not recognized as self-signed.

To fix this, you could change the plugin order, either demote the _openssl_ plugin via `charon.plugins.openssl.load` (see [the docs](https://docs.strongswan.org/docs/5.9/plugins/pluginLoad.html) for details - or don't load it at all if you don't need any algorithms other plugins don't provide e.g. ECDH/ECDSA), or promote the _gmp_ plugin so it's loaded before the _openssl_ plugin. Either way, it should work if the _gmp_ plugin provides the RSA implementation (the _gcrypt_ plugin would work too).

Alternatively, you could adjust the system-wide policy to allow RSA signatures with SHA-1 in OpenSSL via `sudo update-crypto-policies --set DEFAULT:SHA1` (this basically sets `rh-allow-sha1-signatures = yes` in `openssl.cnf`, so this could theoretically also be set for strongSwan only if you get it to load its own `openssl.cnf` e.g. via `OPENSSL_CONF`).

**Comment by stephanhendl** [comment]
Thank you for explaining it deeper. Unfortuantely none of your suggestions to fix the problem with changing the plugin order worked. Neither demote the openssl plugin to the end nor disbling it at all. 

```
[root@vpn4 charon]# swanctl --stats
uptime: 52 minutes, since Sep 06 08:02:55 2023
worker threads: 16 total, 11 idle, working: 4/0/1/0
job queues: 0/0/0/0
jobs scheduled: 0
IKE_SAs: 0 total, 0 half-open
mallinfo: sbrk 2826240, mmap 0, used 912112, free 1914128
loaded plugins: charon-systemd gmp pkcs11 aesni aes des rc2 sha2 sha1 md4 md5 mgf1 random nonce x509 acert pubkey pkcs1 pkcs7 pkcs12 pgp d5519 chapoly xcbc cmac hmac kdf ctr ccm gcm drbg curl sqlite attr kernel-netlink resolve socket-default farp stroke vici updown eap-identi eap-md5 eap-gtc eap-mschapv2 eap-dynamic eap-radius eap-tls eap-ttls eap-peap xauth-generic xauth-eap xauth-pam xauth-noauth dhcp led dup
```

The CA-certificate is still not marked as self-signed. The only configuration that worked was the command `update-crypto-policies --set DEFAULT:SHA1` what enabled SHA-1 again (see [Chapter 3. Using system-wide cryptographic policies](https://access.redhat.com/documentation/de-de/red_hat_enterprise_linux/9/html/security_hardening/using-the-system-wide-cryptographic-policies_security-hardening)). With that I could load openssl before gmp and it worked as well.

```
[root@vpn4 charon]# swanctl --list-certs --flag ca

  subject:  "C=DE, ST=Brandenburg, L=Potsdam, O=Landtag Brandenburg, OU=Verwaltung, CN=Landtag Brandenburg UCS Root CA, E=ca@landtag.brandenburg.de"
  issuer:   "C=DE, ST=Brandenburg, L=Potsdam, O=Landtag Brandenburg, OU=Verwaltung, CN=Landtag Brandenburg UCS Root CA, E=ca@landtag.brandenburg.de"
  validity:  not before Feb 18 16:58:21 2011, ok
             not after  Feb 18 16:58:21 2036, ok (expires in 4548 days)
  serial:    82:a6:9a:fe:9e:59:12:07
  altNames:  ca@landtag.brandenburg.de
  flags:     CA CRLSign self-signed
  authkeyId: 56:3f:82:b6:16:c6:bf:46:18:de:6c:b3:a0:e1:c4:10:52:82:9d:41
  subjkeyId: 56:3f:82:b6:16:c6:bf:46:18:de:6c:b3:a0:e1:c4:10:52:82:9d:41
  pubkey:    RSA 2048 bits
  keyid:     00:c4:c5:97:27:a2:71:e6:31:f9:1d:71:bd:5a:05:7d:df:32:97:0a
  subjkey:   56:3f:82:b6:16:c6:bf:46:18:de:6c:b3:a0:e1:c4:10:52:82:9d:41
```

So in order to be on the safe side in future I should have to think about a new or additional root CA.

**Reply by tobiasbrunner** [reply]
> Unfortuantely none of your suggestions to fix the problem with changing the plugin order worked. Neither demote the openssl plugin to the end nor disbling it at all.

That's weird, especially, if you disabled both the _openssl_ and the _gcrypt_ plugins. In the log with increased log levels for _lib_ you should see a series of `builder L4 CRED_PUBLIC_KEY - RSA of plugin '...'` messages, which should end with `'gmp'` if the plugin order is correct.

> So in order to be on the safe side in future I should have to think about a new or additional root CA.

Yes, I'd definitely recommend issuing a new version of the root CA that's self-signed with a signature that uses SHA-256.
